facilitator/src/transport.rs

```diff
@@ -117,51 +130,102 @@ fn basic_runtime() -> Result<Runtime> {
+                // uploads to fail when Hyper tries to re-use a connection that
+                // has been idle too long. Until this is fixed in Rusoto[3], we
+                // construct our own HTTP request dispatcher whose underlying
+                // hyper::Client is configured to timeout idle connections after
+                // 10 seconds. We could also implement retries on our layer[4].
+                //
+                // [1]: https://docs.rs/hyper/0.13.8/hyper/client/struct.Builder.html#method.pool_idle_timeout
+                // [2]: https://aws.amazon.com/premiumsupport/knowledge-center/s3-socket-connection-timeout-error/
+                // [3]: https://github.com/rusoto/rusoto/issues/1686
+                // [4]: https://github.com/abetterinternet/prio-server/issues/41
+                let mut builder = hyper::Client::builder();
+                builder.pool_idle_timeout(Duration::from_secs(10));
+                let connector = HttpsConnector::new();
+                let http_client = rusoto_core::HttpClient::from_builder(builder, connector);
+
+                if use_ambient_credentials {
+                    // Credentials for authenticating to AWS are automatically
+                    // sourced from environment variables or ~/.aws/credentials.
+                    // https://github.com/rusoto/rusoto/blob/master/AWS-CREDENTIALS.md
+                    let credentials_provider = DefaultCredentialsProvider::new()
+                        .context("failed to create credentials provider")?;
+                    Ok(S3Client::new_with(
+                        http_client,
+                        credentials_provider,
+                        region.clone(),
+                    ))
+                } else {
+                    // When running in GKE, the token used to authenticate to
+                    // AWS S3 is made available via the instance metadata
+                    // service. See terraform/modules/kubernetes/kubernetes.tf
+                    // for discussion.
+                    let token = Client::new()
+                        .get(METADATA_SERVICE_TOKEN_URL)
+                        // The audience query parameter is required by the GKE
+                        // metadata service, but since our role assumption
+                        // policy doesn't check `aud`, it doesn't matter what
+                        // audience we put here (see
+                        // terraform/modules/facilitator/facilitator.tf).
+                        .query(&[("audience", "something")])
+                        .header("Metadata-Flavor", "Google")
+                        .send()
+                        .context("failed to fetch auth token from metadata service")?
+                        .text()
+                        .context("failed to unwrap metadata token")?;
+
+                    let credentials_provider = WebIdentityProvider::new(
+                        Variable::with_value(token),
```

Comment:
It was actually pretty easy to wire up a `Variable::dynamic` that does the GET against the metadata service, so I did.

Reply:
I think I spoke too soon: we can't use `reqwest::blocking::Client` from inside a `Variable::Dynamic` because the `Dynamic`'s closure gets invoked from within a Tokio runtime constructed by Rusoto, and `reqwest::blocking` panics if used in a runtime. We can't use the async `reqwest::Client` because the closure isn't explicitly async, so we can't use `await` in there. I worked around this by using [ureq](https://github.com/algesten/ureq) instead of `reqwest::blocking`, as ureq is purely synchronous Rust and thus safe to use in a `Variable::Dynamic`.